src/Storages/StorageS3.cpp

```diff
@@ -45,100 +45,217 @@ namespace ErrorCodes
+        return false;
+    }
+
+    read_buf = wrapReadBufferWithCompressionMethod(
+        std::make_unique<ReadBufferFromS3>(client, bucket, current_key), chooseCompressionMethod(current_key, compression_hint));
+    auto input_format = FormatFactory::instance().getInput(format, *read_buf, sample_block, context, max_block_size);
+    reader = std::make_shared<InputStreamFromInputFormat>(input_format);
 
-        Chunk generate() override
+    if (columns_desc.hasDefaults())
+        reader = std::make_shared<AddingDefaultsBlockInputStream>(reader, columns_desc, context);
+
+    initialized = false;
+    return true;
+}
+
+String StorageS3Source::getName() const
+{
+    return name;
+}
+
+Chunk StorageS3Source::generate()
+{
+    if (!reader)
+        return {};
+
+    if (!initialized)
+    {
+        reader->readPrefix();
+        initialized = true;
+    }
+
+    if (auto block = reader->read())
+    {
+        auto columns = block.getColumns();
+        UInt64 num_rows = block.rows();
+
+        if (with_path_column)
+            columns.push_back(DataTypeString().createColumnConst(num_rows, file_path)->convertToFullColumnIfConst());
+        if (with_file_column)
         {
-            if (!reader)
-                return {};
-
-            if (!initialized)
-            {
-                reader->readSuffix();
-                initialized = true;
-            }
-
-            if (auto block = reader->read())
-            {
-                auto columns = block.getColumns();
-                UInt64 num_rows = block.rows();
-
-                if (with_path_column)
-                    columns.push_back(DataTypeString().createColumnConst(num_rows, file_path)->convertToFullColumnIfConst());
-                if (with_file_column)
-                {
-                    size_t last_slash_pos = file_path.find_last_of('/');
-                    columns.push_back(DataTypeString().createColumnConst(num_rows, file_path.substr(
-                            last_slash_pos + 1))->convertToFullColumnIfConst());
-                }
-
-                return Chunk(std::move(columns), num_rows);
-            }
-
-            reader.reset();
-
-            return {};
+            size_t last_slash_pos = file_path.find_last_of('/');
+            columns.push_back(DataTypeString().createColumnConst(num_rows, file_path.substr(
+                    last_slash_pos + 1))->convertToFullColumnIfConst());
         }
 
-    private:
-        String name;
-        std::unique_ptr<ReadBuffer> read_buf;
-        BlockInputStreamPtr reader;
-        bool initialized = false;
-        bool with_file_column = false;
-        bool with_path_column = false;
-        String file_path;
-    };
+        return Chunk(std::move(columns), num_rows);
+    }
 
-    class StorageS3BlockOutputStream : public IBlockOutputStream
+    reader->readSuffix();
+    reader.reset();
+    read_buf.reset();
+
+    if (!initialize())
+        return {};
```

Comment:
That's ok, but it is already a little bit difficult to verify.
We can probably create another one iterator which return reader ...

Reply:
Decided not to fix.